# <svg style="float: left; padding-right: 20px; width: 150px" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 610.99 493.35"><title>Harvard Berkman Klein Center Logo</title><path d="M5.31,428.77v5.13H16.82v4.83H5.31v9.34H0V423.94H18.37v4.83Z" fill="#656668"></path><path d="M36.52,448.48c-7.45,0-12.79-5.55-12.79-12.41V436a12.48,12.48,0,0,1,12.85-12.48c7.45,0,12.79,5.55,12.79,12.41V436A12.47,12.47,0,0,1,36.52,448.48ZM43.82,436a7.34,7.34,0,0,0-7.3-7.59,7.22,7.22,0,0,0-7.24,7.52V436a7.33,7.33,0,0,0,7.3,7.58,7.22,7.22,0,0,0,7.24-7.52Z" fill="#656668"></path><path d="M71.6,448.07l-5.17-7.72H62.26v7.72H57V423.94H68c5.68,0,9.1,3,9.1,8V432a7.5,7.5,0,0,1-5.17,7.48l5.89,8.62Zm.1-15.89c0-2.27-1.59-3.44-4.17-3.44H62.26v6.92h5.37c2.59,0,4.07-1.38,4.07-3.41Z" fill="#656668"></path><path d="M97.11,448.07V423.94h5.3v24.13Z" fill="#656668"></path><path d="M128.26,448.07l-11.68-15.34v15.34h-5.24V423.94h4.9l11.3,14.86V423.94h5.24v24.13Z" fill="#656668"></path><path d="M152,428.84v19.23h-5.31V428.84h-7.34v-4.9h20v4.9Z" fill="#656668"></path><path d="M165.91,448.07V423.94H184.1v4.72H171.18v4.9h11.37v4.72H171.18v5.07h13.1v4.72Z" fill="#656668"></path><path d="M206.85,448.07l-5.17-7.72h-4.17v7.72H192.2V423.94h11c5.69,0,9.1,3,9.1,8V432a7.49,7.49,0,0,1-5.17,7.48l5.89,8.62Zm.1-15.89c0-2.27-1.58-3.44-4.17-3.44h-5.27v6.92h5.38c2.58,0,4.06-1.38,4.06-3.41Z" fill="#656668"></path><path d="M237.06,448.07l-11.68-15.34v15.34h-5.24V423.94H225l11.31,14.86V423.94h5.23v24.13Z" fill="#656668"></path><path d="M250.43,448.07V423.94h18.2v4.72H255.7v4.9h11.38v4.72H255.7v5.07h13.1v4.72Z" fill="#656668"></path><path d="M287.37,428.84v19.23h-5.3V428.84h-7.34v-4.9h20v4.9Z" fill="#656668"></path><path d="M330,448.55l-2.8-2.83a10.85,10.85,0,0,1-7.06,2.69c-4.62,0-8.1-2.69-8.1-7,0-3.11,1.65-5.28,4.69-6.65a8.12,8.12,0,0,1-1.69-4.86c0-3.35,2.65-6.38,7.3-6.38,4.11,0,6.79,2.65,6.79,6.1,0,3.27-2.06,5.17-5.06,6.34l3.2,3.21a46.63,46.63,0,0,0,2.34-4.14l3.9,2.14a45.19,45.19,0,0,1-3.17,5.1l3.55,3.55ZM319.64,438a3.61,3.61,0,0,0-2.24,3.21c0,1.72,1.41,2.86,3.38,2.86a5.73,5.73,0,0,0,3.48-1.38Zm2.62-10.33a2.28,2.28,0,0,0-2.34,2.48,4.31,4.31,0,0,0,1.41,3c2.1-.79,3.21-1.69,3.21-3.24A2.12,2.12,0,0,0,322.26,427.67Z" fill="#656668"></path><path d="M361.35,448.41a15.33,15.33,0,0,1-10.24-3.86l3.14-3.75a11.06,11.06,0,0,0,7.2,2.93c2.17,0,3.48-.87,3.48-2.28v-.07c0-1.34-.82-2-4.86-3.07-4.86-1.24-8-2.58-8-7.37v-.07c0-4.38,3.51-7.27,8.44-7.27a13.91,13.91,0,0,1,9,3.07l-2.76,4a11.25,11.25,0,0,0-6.27-2.38c-2,0-3.1.93-3.1,2.1v.07c0,1.58,1,2.1,5.2,3.17,4.9,1.27,7.65,3,7.65,7.24v.06C370.2,445.72,366.55,448.41,361.35,448.41Z" fill="#656668"></path><path d="M389.09,448.48a12.36,12.36,0,0,1-12.78-12.41V436a12.48,12.48,0,0,1,12.85-12.48c7.45,0,12.79,5.55,12.79,12.41V436A12.48,12.48,0,0,1,389.09,448.48ZM396.4,436a7.33,7.33,0,0,0-7.31-7.59,7.21,7.21,0,0,0-7.23,7.52V436a7.33,7.33,0,0,0,7.3,7.58,7.22,7.22,0,0,0,7.24-7.52Z" fill="#656668"></path><path d="M420.69,448.48a12.18,12.18,0,0,1-12.37-12.41V436a12.26,12.26,0,0,1,12.58-12.48,12.48,12.48,0,0,1,9.51,3.72l-3.38,3.9a8.91,8.91,0,0,0-6.17-2.73c-4.06,0-7,3.38-7,7.52V436c0,4.13,2.86,7.58,7,7.58,2.76,0,4.45-1.11,6.34-2.83l3.38,3.41A12.31,12.31,0,0,1,420.69,448.48Z" fill="#656668"></path><path d="M437.53,448.07V423.94h5.31v24.13Z" fill="#656668"></path><path d="M451.77,448.07V423.94H470v4.72H457v4.9h11.37v4.72H457v5.07h13.09v4.72Z" fill="#656668"></path><path d="M488.72,428.84v19.23h-5.31V428.84h-7.34v-4.9h20v4.9Z" fill="#656668"></path><path d="M515.91,438.45v9.62H510.6v-9.51l-9.27-14.62h6.21l5.75,9.69,5.86-9.69h6Z" fill="#656668"></path><path d="M20,493l-2.2-5.41H7.62L5.41,493H0l10.34-24.3h4.89L25.57,493Zm-7.3-17.92-3.21,7.82h6.41Z" fill="#656668"></path><path d="M40.19,473.74V493h-5.3V473.74H27.54v-4.9h20v4.9Z" fill="#656668"></path><path d="M81.69,493v-9.69H71.9V493H66.6V468.84h5.3v9.55h9.79v-9.55H87V493Z" fill="#656668"></path><path d="M113.41,493l-2.2-5.41H101L98.8,493H93.39l10.34-24.3h4.89L119,493Zm-7.3-17.92-3.21,7.82h6.41Z" fill="#656668"></path><path d="M140,493l-5.17-7.72h-4.17V493h-5.31V468.84h11c5.69,0,9.1,3,9.1,8v.06a7.5,7.5,0,0,1-5.17,7.48L146.2,493Zm.1-15.89c0-2.28-1.59-3.44-4.17-3.44h-5.27v6.92H136c2.59,0,4.07-1.38,4.07-3.41Z" fill="#656668"></path><path d="M164.73,493.14H160l-9.75-24.3h5.86l6.31,17,6.3-17h5.72Z" fill="#656668"></path><path d="M196.58,493l-2.21-5.41h-10.2L182,493h-5.41l10.34-24.3h4.89L202.13,493Zm-7.31-17.92-3.2,7.82h6.41Z" fill="#656668"></path><path d="M223.18,493,218,485.25h-4.17V493h-5.31V468.84h11c5.68,0,9.1,3,9.1,8v.06a7.5,7.5,0,0,1-5.18,7.48l5.9,8.62Zm.1-15.89c0-2.28-1.58-3.44-4.17-3.44h-5.27v6.92h5.37c2.59,0,4.07-1.38,4.07-3.41Z" fill="#656668"></path><path d="M245.88,493h-9.41V468.84h9.41c7.58,0,12.82,5.21,12.82,12v.07C258.7,487.7,253.46,493,245.88,493Zm7.27-12.06a6.92,6.92,0,0,0-7.27-7.27h-4.11v14.54h4.11a6.88,6.88,0,0,0,7.27-7.21Z" fill="#656668"></path><path d="M288.48,493.35c-6.48,0-10.44-3.62-10.44-10.72V468.84h5.3v13.65c0,3.93,2,6,5.21,6s5.2-2,5.2-5.79V468.84h5.31v13.62C299.06,489.76,295,493.35,288.48,493.35Z" fill="#656668"></path><path d="M324.55,493l-11.68-15.34V493h-5.24V468.84h4.9l11.3,14.86V468.84h5.24V493Z" fill="#656668"></path><path d="M338,493V468.84h5.31V493Z" fill="#656668"></path><path d="M364.26,493.14h-4.69l-9.75-24.3h5.86l6.31,17,6.3-17H374Z" fill="#656668"></path><path d="M380.42,493V468.84h18.2v4.72H385.69v4.9h11.38v4.72H385.69v5.07h13.1V493Z" fill="#656668"></path><path d="M421.37,493l-5.18-7.72H412V493h-5.3V468.84h11c5.68,0,9.09,3,9.09,8v.06a7.49,7.49,0,0,1-5.17,7.48l5.9,8.62Zm.1-15.89c0-2.28-1.59-3.44-4.17-3.44H412v6.92h5.38c2.59,0,4.07-1.38,4.07-3.41Z" fill="#656668"></path><path d="M443,493.31a15.31,15.31,0,0,1-10.23-3.86l3.13-3.76a11.06,11.06,0,0,0,7.21,2.93c2.17,0,3.48-.86,3.48-2.27v-.07c0-1.34-.83-2-4.86-3.07-4.86-1.24-8-2.58-8-7.37v-.07c0-4.38,3.52-7.27,8.45-7.27a14,14,0,0,1,9,3.07l-2.76,4a11.33,11.33,0,0,0-6.27-2.37c-2,0-3.1.93-3.1,2.1v.07c0,1.58,1,2.1,5.2,3.17,4.89,1.27,7.65,3,7.65,7.24v.06C451.85,490.62,448.2,493.31,443,493.31Z" fill="#656668"></path><path d="M459.54,493V468.84h5.31V493Z" fill="#656668"></path><path d="M484.14,473.74V493h-5.31V473.74H471.5v-4.9h20v4.9Z" fill="#656668"></path><path d="M511.34,483.35V493H506v-9.51l-9.27-14.62H503l5.76,9.69,5.86-9.69h6Z" fill="#656668"></path><path d="M0,263.52H26.21c6.42,0,10.95,1.61,13.95,4.61a11.2,11.2,0,0,1,3.28,8.24v.15c0,5.91-3.35,9.42-7.88,11.39,6.2,2,10.14,5.55,10.14,12.49v.14c0,8.76-7.15,14.09-19.34,14.09H0Zm22.71,20.37c4.3,0,6.71-1.46,6.71-4.45v-.14c0-2.78-2.19-4.39-6.42-4.39H13.8v9Zm2,19.35c4.31,0,6.79-1.68,6.79-4.67v-.15c0-2.77-2.19-4.6-6.94-4.6H13.8v9.42Z"></path><path d="M58.28,263.52H99.39v12.05H72.3v7.74H96.83v11.17H72.3v8.11H99.75v12H58.28Z"></path><path d="M113.49,263.52h24.17c7.81,0,13.21,2.05,16.64,5.55,3,2.93,4.53,6.87,4.53,11.9v.15c0,7.81-4.16,13-10.51,15.7l12.19,17.81H144.16l-10.3-15.48h-6.2v15.48H113.49ZM137,288.06c4.82,0,7.59-2.34,7.59-6.06v-.15c0-4-2.92-6.06-7.66-6.06h-9.27v12.27Z"></path><path d="M171.79,263.52H186v20.74l17.6-20.74h16.79L201,285.5l20,29.13H204l-12.84-19L186,301.42v13.21H171.79Z"></path><path d="M230.65,263.52h15l12.19,19.79L270,263.52h15v51.11H270.88V285.28l-13.08,20h-.29l-13-19.86v29.2H230.65Z"></path><path d="M316.8,263.16h13.65l21.76,51.47H337l-3.73-9.12H313.59l-3.65,9.12H295Zm12.41,31.32L323.52,280l-5.77,14.53Z"></path><path d="M362.29,263.52h13.22l21,27v-27h14v51.11H398.14l-21.83-28v28h-14Z"></path><path d="M0,341.35H14.16v20.74l17.6-20.74H48.55l-19.34,22,20,29.13h-17l-12.85-19-5.19,5.76v13.22H0Z"></path><path d="M58.86,341.35H73v38.7H97.78v12.41H58.86Z"></path><path d="M108.36,341.35h41.1v12H122.38v7.74h24.53v11.17H122.38v8.1h27.45v12H108.36Z"></path><path d="M163.77,341.35H178v51.11H163.77Z"></path><path d="M193.18,341.35H206.4l21,27v-27h14v51.11H229l-21.83-28v28h-14Z"></path><path d="M278.89,367.05v-.14c0-14.9,11.47-26.58,26.94-26.58,10.44,0,17.16,4.38,21.69,10.66l-10.66,8.25c-2.92-3.65-6.28-6-11.17-6-7.16,0-12.19,6.06-12.19,13.51v.15c0,7.66,5,13.65,12.19,13.65,5.33,0,8.47-2.49,11.53-6.21l10.66,7.6c-4.81,6.64-11.31,11.53-22.63,11.53A26,26,0,0,1,278.89,367.05Z"></path><path d="M339.12,341.35h41.1v12H353.13v7.74h24.54v11.17H353.13v8.1h27.46v12H339.12Z"></path><path d="M394.33,341.35h13.21l21,27v-27h14v51.11H430.18l-21.84-28v28h-14Z"></path><path d="M468.61,353.76H453.28V341.35h44.83v12.41H482.77v38.7H468.61Z"></path><path d="M508.76,341.35h41.11v12H522.78v7.74h24.53v11.17H522.78v8.1h27.45v12H508.76Z"></path><path d="M564,341.35h24.17c7.81,0,13.21,2,16.65,5.55,3,2.92,4.52,6.86,4.52,11.9V359c0,7.81-4.16,13-10.51,15.69L611,392.46H594.64L584.35,377h-6.21v15.48H564Zm23.51,24.53c4.82,0,7.6-2.33,7.6-6.06v-.14c0-4-2.92-6.06-7.67-6.06h-9.27v12.26Z"></path><rect x="125.1" width="10" height="171.1" fill="#a51c30"></rect><rect x="286.93" width="10" height="171.1" fill="#a51c30"></rect><polygon points="186.8 171.29 179.73 164.22 227.51 116.44 179.59 68.52 186.66 61.45 241.65 116.44 186.8 171.29" fill="#a51c30"></polygon><polygon points="397.05 171.29 404.12 164.22 356.35 116.44 404.27 68.52 397.19 61.45 342.21 116.44 397.05 171.29" fill="#a51c30"></polygon><polygon points="0.94 149.47 22.48 171.01 76.75 116.75 21.54 61.55 0 83.09 33.66 116.75 0.94 149.47" fill="#a51c30"></polygon></svg> BKC Research Sprint - Spring 2024
### Visualizing Boston's Urban Spaces Using Open Data

## Group 5: Landmarks of Boston
**Team Members: Justin Wei, Gemma Li, Kevin Hu and Krithika Balaraman**
### Exploratory Data Analysis and Visualization Drafts


<hr style="height:2pt">

In [21]:
# Load libraries
import pandas as pd
import textwrap

Dataset: BOSTON BUILDINGS INVENTORY

Source: https://data.boston.gov/dataset/boston-buildings-inventory

Description: 
This dataset pulls from many different data sources to identify individual building characteristics of all buildings in Boston. It also identifies high-potential retrofit options to reduce carbon emissions in multifamily buildings, using the best available data and assumptions from building experts.

In [10]:
# Load main datafile: BOSTON BUILDING INVENTORY
df = pd.read_csv("building_inventory_021020.csv")
df.head()

C:\Users\JustinMingganWei\AppData\Local\Temp\ipykernel_10672\3769625082.py:2: DtypeWarning: Columns (1,40,44,47,51,53) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("building_inventory_021020.csv")


id    pid_long  cm_id    building_typology  \
0  88905         NaN    NaN                  NaN   
1  51137  1600083000    NaN        Single-Family   
2  18045   202560000    NaN        Single-Family   
3  94001  2200577000    NaN  Convention/Assembly   
4  92940  1600058000    NaN                  NaN   

             building_subtypology              use_class     sqft_class  \
0                             NaN                    NaN            NaN   
1        Single-Family, 1915-1949            Residential  < 25,000 sqft   
2         Single-Family, Pre-1915            Residential  < 25,000 sqft   
3  Convention/Assembly, 1950-1979  Public & Quasi-Public  < 25,000 sqft   
4                             NaN                    NaN            NaN   

  year_built_class st_num  st_name  ... insulate_exposed_ducts  \
0              NaN    NaN      NaN  ...                      f   
1        1915-1949      5  EVERETT  ...                      f   
2         Pre-1915     43    ESSEX  ...                      f   
3        1950-1979    525  WESTERN  ...                      f   
4              NaN    NaN      NaN  ...                      f   

   insulate_exposed_pipes  interior_wall_insulation_blow_in  \
0                       f                                 f   
1                       t                                 f   
2                       t                                 f   
3                       f                                 f   
4                       f                                 f   

   exterior_wall_insulation_at_replacement  exterior_wall_insulation  \
0                                        f                         f   
1                                        f                         t   
2                                        f                         f   
3                                        f                         f   
4                                        f                         f   

   interior_wall_insulation_board  insulate_spandrel  asbestos  \
0                               f                  f         f   
1                               t                  f         f   
2                               t                  f         f   
3                               f                  f         f   
4                               f                  f         f   

   seal_elevator_vent_shafts  env_recommendation_count  
0                          f                         0  
1                          f                         4  
2                          f                         3  
3                          f                         0  
4                          f                         0  

[5 rows x 107 columns]

In [17]:
print("This dataset contains", df.shape[0], "entries building information, with", df.shape[1], "attributes for each.") 

This dataset contains 98930 entries building information, with 107 attributes for each.


In [39]:
columns_str = ', '.join(df.columns.tolist())
wrapped_columns = textwrap.fill(columns_str, width=128)
print("Here is a full list of all the columns / attributes:\n\n"+wrapped_columns)

Here is a full list of all the columns / attributes:

id, pid_long, cm_id, building_typology, building_subtypology, use_class, sqft_class, year_built_class, st_num, st_name,
st_name_suf, unit_num, zipcode, yr_built, land_sf, gross_area, living_area, sqft, num_floors, num_bldgs, units_res, units_com,
units_mixed, ptype, ptype_list, lu, lu_list, assessor_category, assessor_description, own_occ, own_occ_list, owner_list,
structure_class, structure_class_list, bldg_styl, bldg_styl_list, r_roof_typ, ext_fin, ext_fin_list, ext_cond_list,
insulation_type, foundation_type, heat_typ, heat_typ_list, heating_system_fuel, ac, ac_list, ac_system_type,
hot_water_system_fuel, electric_panel_info, has_pv, existing_ashp, existing_gshp, existing_shw, historic_district, landmark,
flood, stormwater, last_major_renovation_date, total_site_energy_kbtu, perc_electricity, perc_gas, perc_steam,
onsite_renewable_kwh, census_tract_number, ct_pop_disability, ct_pop_children_under_5, ct_pop_over_65, ct_pop_low_to_

In [118]:
df_description = pd.read_csv("fields-definitions.csv")
df_description['Field Name'] = df_description['Field Name'].str.strip() # Cleansing for extra spaces
df_cat = df_description.Category.unique()
print("Main", len(df_cat), "attribute categories:\n\n", df_cat)

Main 13 attribute categories:

 ['General information' 'Uses' 'Other' 'Building age' 'Building size'
 'Height' 'Number of buildings' 'Number of units' 'Zoning' 'Socioeconomic'
 'Environmental' 'Heat Pump Strategy' 'Envelope Strategy']


With the supplementary data fields descriptions file, we see that there are 13 main categories of attributes. We can use this to group the 107 attributes into smaller groups for easier operations.

The dataset contains all buildings in Boston, but we are only interested in those that are landmarks:

In [84]:
filtered_df = df[df['landmark'] == 't']
print("Number of all landmarks:", len(filtered_df))

Number of all landmarks: 632


Check if there is anything be of our interests in the `Uses` attributes:

In [85]:
df_description[df_description.Category == "Uses"][["Field Name", "Description"]]

Field Name                                        Description
8                  ptype                                 Property type code
9             ptype_list  List of property type codes (where data was ag...
10                    lu                                      Land use code
11               lu_list  List of land use codes (where data was aggrega...
12               own_occ  Whether the property receives a tax exemption ...
13          own_occ_list  List of owner occupany types (where data was a...
37     assessor_category       Property type category associated with ptype
38  assessor_description    Property type description associated with ptype
76             use_class                         General use classification
79     building_typology           General building typology classification
80  building_subtypology  Building typology classification including sub...

In [88]:
filtered_df[df_description[df_description.Category == "Uses"]["Field Name"].tolist()].head()

ptype ptype_list  lu lu_list own_occ own_occ_list     assessor_category  \
3    941.0        941   E       E       N            N  Exempt Property Type   
71   986.0        986   E       E       N            N  Exempt Property Type   
112  102.0    102,995  CD   CD,CM       N          N,Y  Residential Property   
124  995.0        995  CM      CM       N            N  Exempt Property Type   
129  995.0        995  CM      CM       N            N  Exempt Property Type   

     assessor_description                use_class  \
3    Auditorium / Theater    Public & Quasi-Public   
71            Vacant Land                   Vacant   
112     Residential Condo              Residential   
124            Condo Main  Residential (Mixed-Use)   
129            Condo Main              Residential   

               building_typology                       building_subtypology  
3            Convention/Assembly             Convention/Assembly, 1950-1979  
71                        Vacant                                     Vacant  
112  Multi-Family (>= 100 units)  Multi-Family (>= 100 units), 1980-present  
124      Residential (Mixed-Use)      Residential (Mixed-Use), 1980-present  
129                Single-Family                Single-Family, 1980-present

We want to look at what are the `General Use Classification` of these landmarks:

In [72]:
filtered_df.groupby('use_class').size().sort_values(ascending=False)

use_class
Residential                209
Public & Quasi-Public      198
Commercial                 155
Vacant                      35
Residential (Mixed-Use)     34
Industrial                   1
dtype: int64

<div class="alert alert-success" style="color: #333; background-color: #e6e6fa; border-color: #d8bfd8; border-width: 1px; border-radius: 3px; padding: 10px;">

**⭐ Findings:**
- 33% of the landmarks are `Residential`, 31% are `Public & Quasi-Public`, 25% are `Commercial`.

</div>

Fixing defective entry of `Field Name` in the fields descriptions dataset: 
>`KeyError: "['perc_gas numeric'] not in index"`

In [123]:
df_description.loc[df_description['Field Name'] == 'perc_gas numeric', 'Field Name'] = 'perc_gas'

**Data Cleansing**

Checking for missing data by attribute categories:

>'General information' 'Uses' 'Other' 'Building age' 'Building size'
 'Height' 'Number of buildings' 'Number of units' 'Zoning' 'Socioeconomic'
 'Environmental' 'Heat Pump Strategy' 'Envelope Strategy'

In [ ]:
# for cat in df_cat:
#    filtered_df[df_description[df_description.Category == cat]["Field Name"].tolist()].info(verbose=True, show_counts=True)

In [134]:
df_description[df_description.Category == "General information"][["Field Name", "Description"]]

Field Name              Description
0            id                Unique ID
1      pid_long                Parcel ID
2         cm_id            Condo Main ID
3        st_num            Street number
4       st_name              Street name
5   st_name_suf       Street name suffix
6      unit_num      Address unit number
7       zipcode                 Zip code
14   owner_list  List of property owners

In [131]:
filtered_df[df_description[df_description.Category == 'General information']["Field Name"].tolist()].isnull().sum()

id               0
pid_long         0
cm_id          544
st_num          88
st_name         88
st_name_suf     88
unit_num       632
zipcode         88
owner_list       0
dtype: int64

In [135]:
df_description[df_description.Category == "Uses"][["Field Name", "Description"]]

Field Name                                        Description
8                  ptype                                 Property type code
9             ptype_list  List of property type codes (where data was ag...
10                    lu                                      Land use code
11               lu_list  List of land use codes (where data was aggrega...
12               own_occ  Whether the property receives a tax exemption ...
13          own_occ_list  List of owner occupany types (where data was a...
37     assessor_category       Property type category associated with ptype
38  assessor_description    Property type description associated with ptype
76             use_class                         General use classification
79     building_typology           General building typology classification
80  building_subtypology  Building typology classification including sub...

In [132]:
filtered_df[df_description[df_description.Category == 'Uses']["Field Name"].tolist()].isnull().sum()

ptype                     0
ptype_list                0
lu                        0
lu_list                   0
own_occ                   0
own_occ_list              0
assessor_category         0
assessor_description      0
use_class                 0
building_typology       110
building_subtypology    175
dtype: int64

In [136]:
df_description[df_description.Category == "Other"][["Field Name", "Description"]]

Field Name  \
15                     land_sf   
21             structure_class   
22        structure_class_list   
23                   bldg_styl   
24              bldg_styl_list   
25                  r_roof_typ   
26                     ext_fin   
27                ext_fin_list   
28                    heat_typ   
29               heat_typ_list   
30                          ac   
31                     ac_list   
32               ext_cond_list   
41                      has_pv   
42  last_major_renovation_date   
43         heating_system_fuel   
44              ac_system_type   
45       hot_water_system_fuel   
46             insulation_type   
47             foundation_type   
48         electric_panel_info   
49               existing_ashp   
50               existing_gshp   
51                existing_shw   
52      total_site_energy_kbtu   
53            perc_electricity   
54                    perc_gas   
55                  perc_steam   
56        onsite_renewable_kwh   

                                          Description  
15                  Land area of parcel (square feet)  
21  Commercial building structure class (e.g. meta...  
22  List of building stucture classes (where data ...  
23  Building style (decker, row middle, row end, e...  
24  List of building styles (where data was aggreg...  
25  Roof type for residential parcels (e.g. flat, ...  
26         Exterior finish (e.g. brick/stone, stucco)  
27  List of exterior finishes (where data was aggr...  
28                                Heating system type  
29  List of heating system types (where data was a...  
30                       Air conditioning system type  
31  List of air conditioning system types (where d...  
32                        List of exterior conditions  
41              Whether the property has solar panels  
42     The last major renovation date of the property  
43        Heating system and heating fuel information  
44                              AC system information  
45    Hot water system and hot water fuel information  
46                          Description of insulation  
47                          Description of foundation  
48                         Electric panel information  
49  Whether this property already has an air-sourc...  
50  Whether this property already has a ground-sou...  
51  Whether this property already has a solar hot ...  
52                    Total annual energy use in kBTU  
53  Percentage of total annual energy use from ele...  
54     Percentage of total annual energy use from gas  
55   Percentage of total annual energy use from steam  
56  Total annual on-site renewable energy producti...

In [133]:
filtered_df[df_description[df_description.Category == 'Other']["Field Name"].tolist()].isnull().sum()

land_sf                         0
structure_class               181
structure_class_list          181
bldg_styl                     466
bldg_styl_list                466
r_roof_typ                    469
ext_fin                       466
ext_fin_list                  466
heat_typ                      473
heat_typ_list                 476
ac                            476
ac_list                       476
ext_cond_list                 466
has_pv                        622
last_major_renovation_date    352
heating_system_fuel           632
ac_system_type                632
hot_water_system_fuel         631
insulation_type               632
foundation_type               631
electric_panel_info           631
existing_ashp                 632
existing_gshp                 632
existing_shw                  632
total_site_energy_kbtu        562
perc_electricity              562
perc_gas                      576
perc_steam                    617
onsite_renewable_kwh          632
dtype: int64

In [137]:
df_description[df_description.Category == "Building age"][["Field Name", "Description"]]

Field Name                  Description
16          yr_built                   Year built
78  year_built_class  Building age classification

In [138]:
filtered_df[df_description[df_description.Category == 'Building age']["Field Name"].tolist()].isnull().sum()

yr_built            140
year_built_class    140
dtype: int64

In [139]:
df_description[df_description.Category == "Building size"][["Field Name", "Description"]]

Field Name                                        Description
17   gross_area  Building area for commercial parcels (square f...
18  living_area  Building living area for residential parcels (...
19         sqft                        Building area (square feet)
77   sqft_class  Square footage classification based on sqft field

In [140]:
filtered_df[df_description[df_description.Category == 'Building size']["Field Name"].tolist()].isnull().sum()

gross_area      19
living_area     19
sqft           105
sqft_class     136
dtype: int64

In [141]:
df_description[df_description.Category == "Height"][["Field Name", "Description"]]

Field Name       Description
20  num_floors  Number of floors

In [142]:
filtered_df[df_description[df_description.Category == 'Height']["Field Name"].tolist()].isnull().sum()

num_floors    125
dtype: int64

In [143]:
df_description[df_description.Category == "Number of buildings"][["Field Name", "Description"]]

Field Name          Description
33  num_bldgs  Number of buildings

In [144]:
filtered_df[df_description[df_description.Category == 'Number of buildings']["Field Name"].tolist()].isnull().sum()

num_bldgs    41
dtype: int64

In [145]:
df_description[df_description.Category == "Number of units"][["Field Name", "Description"]]

Field Name                  Description
34    units_res  Number of residential units
35    units_com   Number of commercial units
36  units_mixed    Number of mixed-use units

In [146]:
filtered_df[df_description[df_description.Category == 'Number of units']["Field Name"].tolist()].isnull().sum()

units_res      418
units_com      540
units_mixed    544
dtype: int64

In [147]:
df_description[df_description.Category == "Zoning"][["Field Name", "Description"]]

Field Name                                     Description
39  historic_district  Whether the property is in a historic district
40           landmark  Whether the property is in a landmark district

In [148]:
filtered_df[df_description[df_description.Category == 'Zoning']["Field Name"].tolist()].isnull().sum()

historic_district    0
landmark             0
dtype: int64

In [149]:
df_description[df_description.Category == "Socioeconomic"][["Field Name", "Description"]]

Field Name  \
57                  census_tract_number   
58                    ct_pop_disability   
59              ct_pop_children_under_5   
60                       ct_pop_over_65   
61              ct_pop_low_to_no_income   
62   ct_pop_limited_english_proficiency   
63                           ct_pop_poc   
64                   ct_pop_med_illness   
65                   ct_perc_disability   
66             ct_perc_children_under_5   
67                      ct_perc_over_65   
68             ct_perc_low_to_no_income   
69  ct_perc_limited_english_proficiency   
70                          ct_perc_poc   
71                  ct_perc_med_illness   
72          ct_hh_income_200000_or_more   
73        ct_perc_income_200000_or_more   

                                          Description  
57                                Census Tract number  
58  Population in Census Tract of people with disa...  
59       Population in Census Tract under 5 years old  
60       Population in Census Tract over 65 years old  
61  Population in Census Tract with income 100% be...  
62  Population in Census Tract with limited Englis...  
63      Population in Census Tract of people of color  
64  Population in Census Tract of people with medi...  
65  Percentage of population in Census Tract of pe...  
66  Percentage of population in Census Tract under...  
67  Percentage of population in Census Tract over ...  
68  Percentage of population in Census Tract with ...  
69  Percentage of population in Census Tract with ...  
70  Percentage of population in Census Tract of pe...  
71  Percentage of population in Census Tract of pe...  
72  Number of households in Census Tract with inco...  
73  Percentage of households in Census Tract with ...

In [150]:
filtered_df[df_description[df_description.Category == 'Socioeconomic']["Field Name"].tolist()].isnull().sum()

census_tract_number                    12
ct_pop_disability                      12
ct_pop_children_under_5                12
ct_pop_over_65                         12
ct_pop_low_to_no_income                12
ct_pop_limited_english_proficiency     12
ct_pop_poc                             12
ct_pop_med_illness                     12
ct_perc_disability                     12
ct_perc_children_under_5               12
ct_perc_over_65                        12
ct_perc_low_to_no_income               12
ct_perc_limited_english_proficiency    12
ct_perc_poc                            12
ct_perc_med_illness                    12
ct_hh_income_200000_or_more            20
ct_perc_income_200000_or_more          20
dtype: int64

In [151]:
df_description[df_description.Category == "Environmental"][["Field Name", "Description"]]

Field Name                                        Description
74       flood  Whether the property falls within the 1% annua...
75  stormwater  Whether the property falls within the stormwat...

In [152]:
filtered_df[df_description[df_description.Category == 'Environmental']["Field Name"].tolist()].isnull().sum()

flood         0
stormwater    0
dtype: int64

In [153]:
df_description[df_description.Category == "Heat Pump Strategy"][["Field Name", "Description"]]

Field Name                                        Description
81               cashp_only  Whether ducted (central) air source heat pump ...
82   cashp_and_elec_upgrade  Whether ducted (central) air source heat pump ...
83            cashp_outdoor  Whether ducted (central) air source heat pump ...
84               dashp_only  Whether ductless air source heat pump replacem...
85   dashp_and_elec_upgrade  Whether ductless air source heat pump replacem...
86            dashp_outdoor  Whether ductless air source heat pump replacem...
87                gshp_only  Whether ground source heat pump replacement or...
88    gshp_and_elec_upgrade  Whether ground source heat pump replacement or...
89                 vrf_only  Whether variable refrigerant flow replacement ...
90     vrf_and_elec_upgrade  Whether variable refrigerant flow replacement ...
91                awhp_only  Whether air-to-water heat pump replacement or ...
92    awhp_and_elec_upgrade  Whether air-to-water heat pump replacement or ...
93  hp_recommendation_count  The number of heat pump retrofit strategies re...

In [154]:
filtered_df[df_description[df_description.Category == 'Heat Pump Strategy']["Field Name"].tolist()].isnull().sum()

cashp_only                 0
cashp_and_elec_upgrade     0
cashp_outdoor              0
dashp_only                 0
dashp_and_elec_upgrade     0
dashp_outdoor              0
gshp_only                  0
gshp_and_elec_upgrade      0
vrf_only                   0
vrf_and_elec_upgrade       0
awhp_only                  0
awhp_and_elec_upgrade      0
hp_recommendation_count    0
dtype: int64

In [155]:
df_description[df_description.Category == "Envelope Strategy"][["Field Name", "Description"]]

Field Name  \
94                            insulate_attic   
95                  insulate_attic_converted   
96                       ext_roof_insulation   
97                    insulate_exposed_ducts   
98                    insulate_exposed_pipes   
99          interior_wall_insulation_blow_in   
100  exterior_wall_insulation_at_replacement   
101                 exterior_wall_insulation   
102           interior_wall_insulation_board   
103                        insulate_spandrel   
104                                 asbestos   
105                seal_elevator_vent_shafts   
106                 env_recommendation_count   

                                           Description  
94             Whether attic insulation is recommended  
95   Whether insulation is recommended in a convert...  
96     Whether exterior roof insulation is recommended  
97     Whether insulating exposed ducts is recommended  
98     Whether insulating exposed pipes is recommended  
99   Whether blow-in interior wall insulation is re...  
100  Whether exterior wall insulation is recommende...  
101    Whether exterior wall insulation is recommended  
102  Whether board interior wall insulation is reco...  
103        Whether insulating spandrels is recommanded  
104  Whether the property's exterior finish is comp...  
105  Whether sealing elevator vent shafts is recomm...  
106  The number of envelope retrofit strategies rec...

In [156]:
filtered_df[df_description[df_description.Category == 'Envelope Strategy']["Field Name"].tolist()].isnull().sum()

insulate_attic                             0
insulate_attic_converted                   0
ext_roof_insulation                        0
insulate_exposed_ducts                     0
insulate_exposed_pipes                     0
interior_wall_insulation_blow_in           0
exterior_wall_insulation_at_replacement    0
exterior_wall_insulation                   0
interior_wall_insulation_board             0
insulate_spandrel                          0
asbestos                                   0
seal_elevator_vent_shafts                  0
env_recommendation_count                   0
dtype: int64

<div class="alert alert-success" style="color: #333; background-color: #e6e6fa; border-color: #d8bfd8; border-width: 1px; border-radius: 3px; padding: 10px;">

**⭐ Findings:**
- In `General Information` category, we will rely on `st_num`, `st_name`, and `zipcode` in order to locate the respective building on a map. There are 88 entries that are missing the street address information, so we will drop those entries.
- In `Uses` category, there are over 100 entries has missing values in `building_typology` and `building_subtypology`, which are the "General building typology classification" and "Building typology classification including subtypology based on building age". We have not decided what to do with this pair of features, so we are keeping these entries for now.
- In `Other` category, only `land_sf` has no missing data; all other attributes have great missingness and are mostly unuseable. We may consider only keeping the "Land area of parcel (square feet)" and dropping all other columns.
- in `Building age` category, `yr_built` is a key attribute we will be using for our visualization. There are 140 entries are missing "Year built" so we will have to drop those entries.
- in `Building size`, `Height`, `Number of buildings`, `Number of units` categories, there are a few entries missing `sqft`, `sqft_class`, `num_floors`, `num_bldgs`, `units_res`, `units_com`, `units_mixed`. We will keep these for now as we continue exploring.
- in `Socioeconomic` category, all attributes are related to Census Tract, we will set them aside for now as we continue exploring.

</div>

Now dropping select entries with missing data:

In [174]:
cols_to_scan = ['st_num', 'st_name', 'zipcode', 'yr_built']
cols_to_drop = ['bldg_styl',
 'bldg_styl_list',
 'r_roof_typ',
 'ext_fin',
 'ext_fin_list',
 'heat_typ',
 'heat_typ_list',
 'ac',
 'ac_list',
 'ext_cond_list',
 'has_pv',
 'last_major_renovation_date',
 'heating_system_fuel',
 'ac_system_type',
 'hot_water_system_fuel',
 'insulation_type',
 'foundation_type',
 'electric_panel_info',
 'existing_ashp',
 'existing_gshp',
 'existing_shw',
 'total_site_energy_kbtu',
 'perc_electricity',
 'perc_gas',
 'perc_steam',
 'onsite_renewable_kwh']

In [172]:
filtered_df = filtered_df.drop(columns=cols_to_drop)

In [176]:
cleaned_filtered_df = filtered_df.dropna(subset=cols_to_scan, how='any')
cleaned_filtered_df = cleaned_filtered_df.drop(columns=df_description[df_description.Category == 'Socioeconomic']["Field Name"].tolist())

Dataset overview after cleansing:

In [182]:
cleaned_filtered_df.describe()

id  cm_id  unit_num      zipcode     yr_built       land_sf  \
count    404.000000    0.0       0.0   404.000000   404.000000  4.040000e+02   
mean   53269.274752    NaN       NaN  2132.534653  1917.596535  8.639063e+04   
std    29931.118523    NaN       NaN    37.440325    31.783356  5.091808e+05   
min        7.000000    NaN       NaN  2108.000000  1845.000000  0.000000e+00   
25%    25174.250000    NaN       NaN  2111.000000  1899.000000  4.006000e+03   
50%    60248.500000    NaN       NaN  2119.000000  1906.000000  9.548500e+03   
75%    78442.750000    NaN       NaN  2130.000000  1925.000000  2.535450e+04   
max    98792.000000    NaN       NaN  2467.000000  2018.000000  7.397130e+06   

         gross_area   living_area          sqft  num_floors  num_bldgs  \
count  4.040000e+02  4.040000e+02  4.040000e+02  402.000000      404.0   
mean   7.636403e+04  6.616423e+04  7.486486e+04    4.889303        1.0   
std    1.832390e+05  1.643976e+05  1.830248e+05    5.209411        0.0   
min    1.000000e+00  1.000000e+00  1.000000e+00    1.000000        1.0   
25%    6.145750e+03  4.321500e+03  5.043000e+03    2.000000        1.0   
50%    1.721000e+04  1.364500e+04  1.549000e+04    4.000000        1.0   
75%    6.604075e+04  5.308875e+04  6.520075e+04    5.000000        1.0   
max    1.502842e+06  1.409965e+06  1.502842e+06   42.000000        1.0   

        units_res  units_com  units_mixed       ptype  \
count  125.000000   4.000000          0.0  404.000000   
mean    19.584000   5.000000          NaN  431.000000   
std     38.648787   1.154701          NaN  374.490642   
min      1.000000   4.000000          NaN   13.000000   
25%      2.000000   4.000000          NaN  105.000000   
50%      3.000000   5.000000          NaN  327.000000   
75%     24.000000   6.000000          NaN  906.000000   
max    273.000000   6.000000          NaN  986.000000   

       hp_recommendation_count  env_recommendation_count  
count               404.000000                 404.00000  
mean                  0.131188                   0.75495  
std                   0.410915                   1.14543  
min                   0.000000                   0.00000  
25%                   0.000000                   0.00000  
50%                   0.000000                   0.00000  
75%                   0.000000                   1.00000  
max                   3.000000                   3.00000

<div class="alert alert-success" style="color: #333; background-color: #e6e6fa; border-color: #d8bfd8; border-width: 1px; border-radius: 3px; padding: 10px;">

**⭐ Findings:**
- For `yr_built`, the oldest landmark was built in 1845 (almost 200 years ago!); the newest landmark was built in 2018.
- For `land_sf`, the largest landmark sits on a land of 7397130 sqft, that's 170 acres! The largest `sqft`, the actual square footage is 1502842 sqft, that's 34.5 acres.
- The tallest landmarks has 42 floors according to `num_floors`.

</div>

**Getting Building Geographic Coordinate**

In order to be able to plot these landmarks buildings accurately on map, we need to use their corrosponding longitude and latitude coordinates. The dataset doesn't come with buildings' coordinates, so we will have to get the coordinates based on their street addresses.

<div class="alert alert-success" style="color: #333; background-color: #e6e6fa; border-color: #d8bfd8; border-width: 1px; border-radius: 3px; padding: 10px;">

**⭐ Next Steps:**
- We will need find a way to get geographic coordinated for all landmarks. There should be related dataset or API we can utilize from the City of Boston. Worst case scenario, we will use street address to inference respective coordinates.

</div>